src/node_file.cc

```diff
@@ -102,6 +103,50 @@ using v8::Value;
 
 #define GET_OFFSET(a) ((a)->IsNumber() ? (a)->IntegerValue() : -1)
 
+FD::FD(Environment* env, int fd)
+    : AsyncWrap(env,
+                env->fd_constructor_template()
+                    ->NewInstance(env->context()).ToLocalChecked(),
+                AsyncWrap::PROVIDER_FD), fd_(fd) {
+  MakeWeak<FD>(this);
+  v8::PropertyAttribute attr =
+      static_cast<v8::PropertyAttribute>(v8::ReadOnly | v8::DontDelete);
+  object()->DefineOwnProperty(env->context(),
+                              FIXED_ONE_BYTE_STRING(env->isolate(), "fd"),
+                              Integer::New(env->isolate(), fd),
+                              attr).FromJust();
+}
+
+FD::~FD() {
+  Close();
+  CHECK(closed_);
+  CHECK(persistent().IsEmpty());
+  if (!object().IsEmpty())
+    ClearWrap(object());
+  persistent().Reset();
+}
+
+bool FD::Close() {
+  if (closed_) return false;
+  closed_ = true;
+  uv_fs_t req;
+
+  // TODO(jasnell): For now, this is swallowing any errors that may be
+  // occurring while cleaning up. That is not necessarily a good thing,
+  // but as this is typically happening during GC, there's not a lot we
+  // can reasonably do here.
+  int ret = uv_fs_close(env()->event_loop(), &req, fd_, nullptr);
```

Comment:
Take note: this uses a sync close. If we decide that it's necessary to use an async close, we can switch to that.

There is an open question about what to do with the result. We can use the new `env->SetImmediate()` to schedule a callback out to javascript land after the close attempt in order to report any errors or to give the equivalent to `fs.close(fd, callback)`, but given that the primary use case is for this to close on garbage collection, I did not consider adding that mechanism to be critical.

Reply:
If there is an error, will the file descriptor be leaked? Can we assure that we have no leak before swallowing?
This mechanism implies that we are cleaning up after ourselves whatever happens.

IMHO it should go to ‘uncaughtException’ if it is failing, and he OS fd is still there.